src/bin/pgcopydb/ld_transform.c

```diff
@@ -1170,20 +1185,137 @@ parseMessage(StreamContext *privateContext, char *message, JSON_Value *json)
- * transaction.
- *
- * There are two ways to append a statement to an existing transaction:
- *
- *  1. it's a new statement altogether, we just append to the linked-list
+ * coalesceLogicalTransactionStatement appends a new entry to an existing tuple
+ * array created during the last INSERT statement in a logical transaction.
  *
- *  2. it's the same statement as the previous one, we only add an entry to the
- *     already existing tuple array created on the previous statement
+ * This functionality enables the generation of multi-values INSERT or COPY
+ * commands, enhancing efficiency.
  *
- * This allows to then generate multi-values insert commands, for instance.
+ * Important: Before invoking this function, ensure that validation is performed
+ * using canCoalesceLogicalTransactionStatement.
+ */
+static bool
+coalesceLogicalTransactionStatement(LogicalTransaction *txn,
+									LogicalTransactionStatement *new)
+{
+	LogicalTransactionStatement *last = txn->last;
+
+	LogicalMessageValuesArray *lastValuesArray = &(last->stmt.insert.new.array->values);
+	LogicalMessageValuesArray *newValuesArray = &(new->stmt.insert.new.array->values);
+
+	int capacity = lastValuesArray->capacity;
+	LogicalMessageValues *array = lastValuesArray->array;
+	if (capacity < (lastValuesArray->count + 1))
+	{
+		/* double the capacity to avoid excessive realloc */
+		capacity *= 2;
+		array = (LogicalMessageValues *) realloc(array, sizeof(LogicalMessageValues) *
+												 capacity);
+		if (array == NULL)
+		{
+			log_error(ALLOCATION_FAILED_ERROR);
+			return false;
+		}
+
+		lastValuesArray->capacity = capacity;
+		lastValuesArray->array = array;
+	}
+
+	/* Move the new value from the 'newValuesArray' to the 'lastValuesArray' of
+	 * the existing statement. Additionally, set the count of the 'newValuesArray'
+	 * to 0 to prevent it from being deallocated by FreeLogicalMessageTupleArray,
+	 * as it has been moved to the 'lastValuesArray'.
+	 */
+	lastValuesArray->array[lastValuesArray->count++] = newValuesArray->array[0];
+	newValuesArray->count = 0;
+
+	/* Deallocate the tuple and the new statement */
+	FreeLogicalMessageTupleArray(&(new->stmt.insert.new));
+	free(new);
+
+	++txn->count;
+	return true;
+}
+
+
+/*
+ * canCoalesceLogicalTransactionStatement checks the new statement is
+ * same as the last statement in the txn by comparing the relation name,
+ * column count and column names.
  *
- * TODO: at the moment we don't pack several statements that look alike into
- * the same one.
+ * This acts as a validation function for coalesceLogicalTransactionStatement.
+ */
+static bool
+canCoalesceLogicalTransactionStatement(LogicalTransaction *txn,
+									   LogicalTransactionStatement *new)
+{
+	/* LogicalTransaction is empty when last is NULL */
+	if (txn == NULL || txn->last == NULL || new == NULL)
+	{
+		return false;
+	}
+
+	LogicalTransactionStatement *last = txn->last;
+
+	/* We support coalesce only for INSERTS */
```

Comment:
We could support it also for UPDATE and DELETE, it's just that we don't at the moment. 

For UPDATE statements, we should start with multiple statement sharing the same SET clause (columns *and* values) and with only the WHERE clause being different, then we can have an UPDATE FROM statement (UPDATE with a join).

For DELETE statement, same idea, we could delete more than one statement at a time from the same table.

Reply:
Sure, I can do it in a separate PR.

Reply:
Well if you prefer to do it here, that's fine too. There is value in finishing up the current PR though ;-)